File: app/src/main/java/za/co/dvt/android/showcase/presentation/listapps/ListAppsFragment.java
package za.co.dvt.android.showcase.presentation.listapps;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.google.firebase.database.FirebaseDatabase;

import java.util.List;

import za.co.dvt.android.showcase.R;
import za.co.dvt.android.showcase.model.AppModel;
import za.co.dvt.android.showcase.repository.impl.FirebaseAppRepository;

/**
 * @author rebeccafranks
 * @since 2017/04/13
 */

public class ListAppsFragment extends Fragment implements ListAppsContract.View {

    private ListAppsContract.Presenter listAppsPresenter;
    private RecyclerView recyclerView;
    private AppAdapter recyclerViewAdapter;
    private ProgressBar progressBar;
    private TextView errorTextView;

    @Nullable
    @Override
    public View onCreateView(final LayoutInflater inflater, @Nullable final ViewGroup container,
                             @Nullable final Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_list_apps, container, false);
        setupRecyclerView(v);
        setupToolbar(v);
        setupErrorViews(v);
        setupPresenter();

        return v;
    }

    private void setupErrorViews(View v) {
        progressBar = (ProgressBar) v.findViewById(R.id.progress_bar);
        errorTextView = (TextView) v.findViewById(R.id.text_view_error_msg);
    }

    @Override
    public void onDestroyView() {
        listAppsPresenter.detachView();
        super.onDestroyView();
    }

    private void setupToolbar(View v) {
        Toolbar toolbar = (Toolbar) v.findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.list_apps_title);
    }

    private void setupPresenter() {
        listAppsPresenter = new ListAppsPresenter(
                new FirebaseAppRepository(FirebaseDatabase.getInstance())); //TODO Extract into injection classes
        listAppsPresenter.attachView(this);
        listAppsPresenter.loadApps();
    }

    private void setupRecyclerView(final View v) {
        recyclerView = (RecyclerView) v.findViewById(R.id.recycler_view_apps);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);
        recyclerViewAdapter = new AppAdapter(null, getContext());
        recyclerView.setAdapter(recyclerViewAdapter);
        final DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
                layoutManager.getOrientation());
        recyclerView.addItemDecoration(dividerItemDecoration);
    }

    @Override
    public void showApps(final List<AppModel> appModelList) {
        recyclerView.setVisibility(View.VISIBLE);
        recyclerViewAdapter.setItems(appModelList);
    }

    @Override
    public void showNoInternetError() {
        showError(getString(R.string.load_apps_no_internet));
    }

    private void showError(String error) {
        errorTextView.setVisibility(View.VISIBLE);
        errorTextView.setText(error);
    }

    @Override
    public void showGenericError() {
        showError(getString(R.string.load_apps_generic_error));
    }

    @Override
    public void showLoadingIndicator() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoadingIndicator() {
        progressBar.setVisibility(View.INVISIBLE);
    }

    @Override
    public void showNoApps() {
        showError(getString(R.string.load_apps_no_apps));
    }

    public static ListAppsFragment newInstance() {
        return new ListAppsFragment();
    }
}

File: app/src/main/java/za/co/dvt/android/showcase/ui/listapps/ListAppsFragment.kt
package za.co.dvt.android.showcase.ui.listapps

import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import io.reactivex.MaybeObserver
import io.reactivex.disposables.Disposable
import timber.log.Timber
import za.co.dvt.android.showcase.R
import za.co.dvt.android.showcase.ShowcaseApplication
import za.co.dvt.android.showcase.injection.ShowcaseFactory
import za.co.dvt.android.showcase.model.AppModel
import za.co.dvt.android.showcase.ui.viewapp.ViewAppActivity

/**
 * @author rebeccafranks
 * *
 * @since 2017/04/13
 */

class ListAppsFragment : Fragment(), ListAppsNavigator {

    private lateinit var recyclerView: RecyclerView
    private lateinit var recyclerViewAdapter: AppAdapter
    private lateinit var progressBar: ProgressBar
    private lateinit var errorTextView: TextView
    private lateinit var listAppsViewModel: ListAppsViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_list_apps, container, false)
        setupRecyclerView(v)
        setupToolbar(v)
        setupErrorViews(v)
        setupViewModel()
        return v
    }

    private fun setupViewModel() {
        listAppsViewModel = ViewModelProviders.of(this,
                ShowcaseFactory(activity.application as ShowcaseApplication))
                .get(ListAppsViewModel::class.java)
        listAppsViewModel.getAppList().subscribe(object : MaybeObserver<List<AppModel>> {
            override fun onSubscribe(d: Disposable?) {
            }

            override fun onSuccess(items: List<AppModel>) {
                hideLoadingIndicator()
                showApps(items)
            }

            override fun onComplete() {

            }

            override fun onError(e: Throwable?) {
                Timber.e("Exception getting apps:", e)
            }
        })
    }

    private fun setupErrorViews(v: View) {
        progressBar = v.findViewById(R.id.progress_bar) as ProgressBar
        errorTextView = v.findViewById(R.id.text_view_error_msg) as TextView
    }

    private fun setupToolbar(v: View) {
        val toolbar = v.findViewById(R.id.toolbar) as Toolbar
        toolbar.setTitle(R.string.list_apps_title)
    }

    private fun setupRecyclerView(v: View) {
        recyclerView = v.findViewById(R.id.recycler_view_apps) as RecyclerView
        val layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerView.layoutManager = layoutManager
        recyclerViewAdapter = AppAdapter(ArrayList(), context, this)
        recyclerView.adapter = recyclerViewAdapter
        val dividerItemDecoration = DividerItemDecoration(recyclerView.context,
                layoutManager.orientation)
        recyclerView.addItemDecoration(dividerItemDecoration)
    }

    fun showApps(appModelList: List<AppModel>) {
        recyclerView.visibility = View.VISIBLE
        recyclerViewAdapter.setItems(appModelList)
    }

    fun showNoInternetError() {
        showError(getString(R.string.load_apps_no_internet))
    }

    private fun showError(error: String) {
        errorTextView.visibility = View.VISIBLE
        errorTextView.text = error
    }

    fun showGenericError() {
        showError(getString(R.string.load_apps_generic_error))
    }

    fun showLoadingIndicator() {
        progressBar.visibility = View.VISIBLE
    }

    fun hideLoadingIndicator() {
        progressBar.visibility = View.INVISIBLE
    }

    fun showNoApps() {
        showError(getString(R.string.load_apps_no_apps))
    }

    override fun onAppClick(app: AppModel) {
        val bundle = Bundle()
        bundle.putString(ViewAppActivity.ARG_APP_ID, app.id)

        startActivity(Intent(activity, ViewAppActivity::class.java), bundle)
    }

    companion object {

        fun newInstance(): ListAppsFragment {
            return ListAppsFragment()
        }
    }
}
